relation_engine_server/utils/json_validation.py

```diff
@@ -60,10 +67,19 @@ def get_schema_validator(schema=None, schema_file=None, validate_at=""):
     # get the appropriate location in the schema
     validation_schema = resolve_pointer(schema, validate_at)
 
-    return Validator(validation_schema, format_checker=FormatChecker())
+    if schema_file:
+        resolver = ExtendedRefResolver(schema_file, schema)
```

Comment:
If this is the call that sets the base URI for the schemas, then wouldn't it make more sense to prepend the `file://` protocol to `schema_file` here?

Reply:
that won't work for all ref resolution as `resolve_remote` (the overridden method from the RefResolver) works on `$ref` values, which is where the `file://` protocol needs to be added. Before calling `resolve_remote`, relative references like `../file.json` are made absolute by comparing them with the base URI, but the scheme (`file://`) doesn't get transferred over from the base URI. Thus there's still the problem of telling the RefResolver how to deal with URIs that don't have a scheme.

e.g. in the schema:

```
  $ref: ../file.yaml
```

`resolve_remote` will get `../file.yaml` as an argument

Reply:
Reread your comment again.. my understanding now is:
* You want to overwrite resolve_remote to handle both json & yaml
* Resolve remote does not get the base uri protocol properly passed through

Reply:
Yes, that is correct. Relative paths are resolved against the base uri, but the protocol is not passed through.

Reply:
Okay, I updated my example to extend RefResolver to handle yaml. It still works fine with the given `file://` base URI. So I'm confused

Reply:
In my example, `file://` is passed through to the `resolve_remote` method. Here is what gets printed for me:

`resolve_remote uri is file:///home/j/tmp/jsonschema-resolver/resolver-test/ref1.json`

`resolve_remote uri is file:///home/j/tmp/jsonschema-resolver/resolver-test/ref2.yaml`

Reply:
It may be that your example is too simple. Checkout this branch and make the following edits:

c. line 70
```
    if schema_file:
        # ADD LINE
        schema_file = 'file://' + schema_file
        resolver = ExtendedRefResolver(schema_file, schema)
    else:
        resolver = ExtendedRefResolver.from_schema(schema)
```

and further down in the file:

c. line 143

```
    def resolve_remote(self, uri):

        print({'base_uri': self.base_uri, 'current uri': uri})

        scheme = urlsplit(uri).scheme
        # if there's no scheme, it's a local file, so prefix it with "file://"
#         if scheme == '':
#             uri = 'file://' + uri
```

then run the tests

Reply:
relevant test output:
```======================================================================
ERROR: test_schema_references (test_json_validation.TestJsonValidation)
Ensure referenced schemas, including those written in yaml, can be accessed.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 774, in resolve_from_url
    document = self.store[url]
  File "/usr/local/lib/python3.7/site-packages/jsonschema/_utils.py", line 22, in __getitem__
    return self.store[self.normalize(uri)]
KeyError: '/app/relation_engine_server/test/data/schema_refs/level_1/edge_type.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 777, in resolve_from_url
    document = self.resolve_remote(url)
  File "/app/relation_engine_server/utils/json_validation.py", line 160, in resolve_remote
    with urlopen(uri) as url:
  File "/usr/local/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.7/urllib/request.py", line 510, in open
    req = Request(fullurl, data)
  File "/usr/local/lib/python3.7/urllib/request.py", line 328, in __init__
    self.full_url = url
  File "/usr/local/lib/python3.7/urllib/request.py", line 354, in full_url
    self._parse()
  File "/usr/local/lib/python3.7/urllib/request.py", line 383, in _parse
    raise ValueError("unknown url type: %r" % self.full_url)
ValueError: unknown url type: '/app/relation_engine_server/test/data/schema_refs/level_1/edge_type.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/relation_engine_server/test/test_json_validation.py", line 363, in test_schema_references
    data=valid_edge_data,
  File "/app/relation_engine_server/utils/json_validation.py", line 116, in run_validator
    if validator.is_valid(data):
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 362, in is_valid
    error = next(self.iter_errors(instance, _schema), None)
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 328, in iter_errors
    for error in errors:
  File "/app/relation_engine_server/utils/json_validation.py", line 33, in set_defaults
    for error in validate_properties(validator, properties, instance, schema):
  File "/usr/local/lib/python3.7/site-packages/jsonschema/_validators.py", line 286, in properties
    schema_path=property,
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 344, in descend
    for error in self.iter_errors(instance, schema):
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 328, in iter_errors
    for error in errors:
  File "/usr/local/lib/python3.7/site-packages/jsonschema/_validators.py", line 259, in ref
    scope, resolved = validator.resolver.resolve(ref)
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 766, in resolve
    return url, self._remote_cache(url)
  File "/usr/local/lib/python3.7/site-packages/jsonschema/validators.py", line 779, in resolve_from_url
    raise exceptions.RefResolutionError(exc)
jsonschema.exceptions.RefResolutionError: unknown url type: '/app/relation_engine_server/test/data/schema_refs/level_1/edge_type.json'

----------------------------------------------------------------------
```
i.e. you still get the paths without the protocol

Reply:
Okay, I will merge to not hold you up but want to look at this a little bit more